pkg/receive/handler.go

```diff
@@ -568,31 +590,51 @@ func (h *Handler) forward(ctx context.Context, tenant string, r replica, wreq *p
 		}
 		key := endpointReplica{endpoint: endpoint, replica: r}
 		if _, ok := wreqs[key]; !ok {
-			wreqs[key] = &prompb.WriteRequest{}
+			wreqs[key] = &writeRequest{
+				batch:   batchId,
+				request: &prompb.WriteRequest{},
+			}
+			batchId++
 		}
 		wr := wreqs[key]
-		wr.Timeseries = append(wr.Timeseries, wreq.Timeseries[i])
+		wr.request.Timeseries = append(wr.request.Timeseries, wreq.Timeseries[i])
```

Comment:
@fpetkovski I was reading this over and over, comparing it to the logic in #5910 to try to figure out what could be different and I think that the `batch` member in the `writeRequest` struct is unnecessary. 

**Why?** 

Because the `batchId` is incremented only when a series is hashed to an `(endpoint, replicationIndex)` (aka `endpointReplica`) that isn't in `wreqs` yet. Each `endpointReplica` and `batchId` match 1:1. Thus we could replace `batchId` with `endpointReplica`. It's like if `batchId` was a reference or shorthand to `endpointReplica`

Do you think my reasoning is correct? Or am I missing something?

Reply:
Hey Douglas, yes I think you're right. However this is also not good enough and I have changed the algorithm to track quorum for each individual time-series inside a write request.